File: funcionesAux.py
#frontera = Devuelve #frontera de la clique ingresada.
def frontera(grafo, n, clique):
	res = 0
	for i in range(len(clique)):
		res += len(nodosAdyacentes(grafo, n, clique[i])) - (len(clique) - 1)

	return res


#formaClique = True si la clique y el nodo ingresado forman una nueva clique.
def formaClique(grafo, clique, nodo):
	#Grafo = Matriz, clique = Lista de nodos, nodo = Nodo.

	res = True
	i = 0
	
	while i < len(clique) and res:

		nodoClique = clique[i]
		res = grafo[nodo][nodoClique]

		i += 1

	return res
				

#nodosAdyacentes = Devuelve los nodos adyacentes al nodo ingresado, en forma de lista.
def nodosAdyacentes(grafo, n, nodo):
	#Grafo = Matriz, n = #Nodos, nodo = Nodo.

	res = []
	for i in range(n):
		if grafo[nodo][i]:
			res.append(i)

	return res


#dameCliques = Devuelve todas las cliques formadas con los nodos ingresados.
def dameCliques(grafo, nodos):
	#Grafo = Matriz, nodos = Lista de nodos.

	cliques=[]
	cantNodos=len(nodos)

	if cantNodos>1:
		cliques+=dameCliques(grafo, nodos[1:])
		elemento=[nodos[0]]
		for sub in cliques[:]:
			
			if formaClique(grafo, sub, elemento[0]):
				cliques.append(sub+elemento)

		cliques.append(elemento)
	else:
		cliques+=[nodos]

	return cliques

#cliquesDeNodo = Devuelve todas las cliques formadas por nodo
def cliquesDeNodo(grafo, nodo, adyacentes):
	#Grafo = Matriz, nodo = Nodo, adyacentes = Lista de nodos adyacentes a nodo.

	cliques=dameCliques(grafo, adyacentes)


	for clique in cliques:
		clique.append(nodo)

	cliques.append([nodo])


  	return cliques



#CMFnodo = Devuelve la frontera maxima de las cliques que forma el nodo ingresado
def CMFnodo(grafo, n, nodo):
	#Grafo = Matriz, n = #Nodos, nodo = Nodo.

	adyacentes = nodosAdyacentes(grafo, n, nodo)

	cliques = cliquesDeNodo(grafo, nodo, adyacentes)

	fronteras = []

	for clique in cliques:
		fronteras.append(frontera(grafo, n, clique))

	return max(fronteras)

File: algoritmos/funcionesAux.java
package algoritmos;
import java.util.*;
import java.util.Random;

public class funcionesAux
{
    //frontera = Devuelve #frontera de la clique ingresada.
    public static int frontera(boolean[][] grafo, int n, List<Integer> clique)
    {
        //int res = 0;
        Set<Integer> res = new HashSet<Integer>();
        for(int i=0;i<clique.size();i++)
        {
            //res = res + 
            res.addAll((Collection<Integer>)nodosAdyacentes(grafo, n, clique.get(i)));//.size() - (clique.size() - 1);
        }
        int resultado = res.size();
        if (clique.size() != 1)
        	resultado = resultado - clique.size();
        return resultado;
    }


    //nodosAdyacentes = Devuelve los nodos adyacentes al nodo ingresado, en forma de array.
    public static List<Integer> nodosAdyacentes(boolean[][] grafo, int n, int nodo)
    {
        List<Integer> listAdy = new ArrayList<Integer>();

        for(int i=0;i<n;i++)
        {
            if(grafo[nodo][i])
            {
                listAdy.add(i);
            }
        }

        return listAdy;
    }



    //formaClique = True si la clique y el nodo ingresado forman una nueva clique.
    public static boolean formaClique(boolean[][] grafo, List<Integer> clique, int nodo)
        {
            boolean res = true;
            int i = 0;
            while(i < clique.size() && res)
            {
                res = grafo[nodo][clique.get(i)];
                i++;
            }
            return res;
        }


    //dameCliques = Devuelve todas las cliques formadas con los nodos ingresados.
    public static List<List<Integer>> dameCliques(boolean[][] grafo, List<Integer> nodos)
    {
        List<List<Integer>> cliques = new ArrayList<List<Integer>>();


        if (nodos.size()>1)
        {

            cliques.addAll(dameCliques(grafo, nodos.subList(1,nodos.size())));
            List<Integer> elemento = new ArrayList<Integer>();
            elemento.add(nodos.get(0));

            for(int i=0;i<cliques.size();i++)
            {
                if(formaClique(grafo, cliques.get(i), elemento.get(0)))
                {   
                    List<Integer> aux = new ArrayList<Integer>();
                   
                    aux.add(elemento.get(0));
                    aux.addAll(cliques.get(i));

                    cliques.add(aux);
                }
            }
            cliques.add(elemento);
        }
        else
        {
            cliques.add(nodos);    
        }

        return cliques;
    }



    //cliquesDeNodo = Devuelve todas las cliques formadas por nodo
    public static List<List<Integer>> cliquesDeNodo(boolean[][] grafo, int n, int nodo)
    {   

        List<Integer> adyacentes = nodosAdyacentes(grafo, n, nodo);

        List<List<Integer>> cliques = dameCliques(grafo, adyacentes);

        for(int i=0;i<cliques.size();i++)
        {
            cliques.get(i).add(nodo);
        }

        List<Integer> nodoList = new ArrayList<Integer>();
        nodoList.add(nodo);
        cliques.add(nodoList);

        return cliques;
    }



    //CMFnodo = Devuelve la frontera maxima de las cliques que forma el nodo ingresado
    public static int CMFnodo(boolean[][] grafo, int n, int nodo)
    {
        List<Integer> adyacentes = nodosAdyacentes(grafo, n, nodo);

        List<List<Integer>> cliques = cliquesDeNodo(grafo, n, nodo);

        int fronteraMax = 0;

        for(int i=0;i<cliques.size();i++)
        {
            int fronteraAux = frontera(grafo, n, cliques.get(i));
            if(fronteraAux > fronteraMax)
            {
                fronteraMax = fronteraAux;
            }
        }

        return fronteraMax;
    }  


    public static int nodoMayorGrado(boolean[][] grafo, int n)
    {   
        int mayorGrado = 0;
        Integer nodoMayorGrado = null;

        for(int nodo=0;nodo<n;nodo++)
        {
            int gradoNodo = funcionesAux.nodosAdyacentes(grafo, n, nodo).size();

            if( gradoNodo > mayorGrado)
            {
                mayorGrado = gradoNodo;
                nodoMayorGrado = nodo;
            }
        }    

        return nodoMayorGrado;
    }







    public static void printGrafo(boolean[][] grafo)
    {   
        int n = grafo[0].length;

        System.out.println("Grafo: ");
        for(int i=0;i<n;i++)
        {
            System.out.println(i + Arrays.toString(grafo[i]).replaceAll("true", "*").replaceAll("false", " "));
        }
        System.out.println(" ");
    }

    
    public static int getNodoIncial(boolean[][] grafo, int n, Inicio inicio) {
    	int nodoInicial = 0;
    	
    	switch (inicio) {
    		case NODO_MAYOR_GRADO:
    			nodoInicial = funcionesAux.nodoMayorGrado(grafo, n); break;
    		case NODO_ALEATORIO:
    			nodoInicial = (int) (Math.random() * n);
    	}
    	return nodoInicial;
    }
    
    

}
